File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode dfs(TreeNode root){
		if(root==null) return root;
        
		TreeNode leftTail = dfs(root.left);
		TreeNode rightTail = dfs(root.right);
        
		if(leftTail!=null){
			leftTail.right = root.right;
			root.right = root.left;
			root.left = null;
		}
        
		if(rightTail!=null)
			return rightTail;
		else if(leftTail!=null)
			return leftTail;
		return root;
	}
    public void flatten(TreeNode root) {
        dfs(root);
    }
}

File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def dfs(self,root):
        if not root: return None
        
        left = self.dfs(root.left)
        right = self.dfs(root.right)
        
        if left:
            left.right = root.right
            root.right = root.left
            root.left = None
            
        if right: return right 
        if left: return left
               
        
        return root

    
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        self.dfs(root)
        